### DISCUSSION TITLE
Using Input Multiplexer (IMUX) Registers (IMRs) on Versal Premium Devices

### DISCUSSION THREAD

**Discussion by oshears** [Xilinx/RapidWright#848] [Q&A] [answered]
Hi RapidWright Team,

I'm currently in the process of using RapidWright to create and implement a custom design for Versal. With the Versal architecture, AMD Xilinx has recently added input multiplexer registers to all of their slices (called IMRs). In the image below I've highlighted where these IMR BELs are in relation to the other elements in a SLICEL site.

![Vivado Versal IMRs](https://github.com/Xilinx/RapidWright/assets/2807080/1e4b0e9d-bf7d-451b-b23f-51560f758355)

I'd like to (1) use RapidWright to create registers and place them on these IMR BELs, followed by (2) the creation of the Versal device image in Vivado. While I am able to do the first of these aforementioned steps, when I write a checkpoint using RapidWright and attempt to call `write_device_image` in Vivado, I get the following DRC error:

```tcl
ERROR: [DRC PDRC-254] SLICE_IMR_FF_use_restricted: Unsupported use of SLICE_X48Y0 site data IMR bel AX_IMR for cell FF_AX_IMR. The register assignment should be changed to a non-IMR FF bel.
```

I am wondering if anyone on your team has been able to successfully write the device image / bitstream for a Versal device that uses these IMR registers? It seems that Vivado is a bit restrictive in their usage and does not make it clear how to instantiate them otherwise. Is there a way that RapidWright can be used to create a design that supports instantiation on these BELs?

**Comment by clavin-xlnx** [comment]
It looks like you've had a chance to post in the main AMD support forums: https://support.xilinx.com/s/question/0D54U00007PfnLfSAJ/how-to-force-the-use-of-imux-registers-imr-in-a-versal-design?language=en_US with no official answer regarding these IMR resources.  We have not attempted to use these resources as they are not supported.  

Would you be able to elaborate on why you are interested in using them?  Perhaps if we can better understand your motivation, we can help you more appropriately.

**Reply by oshears** [reply]
Hi Chris!

For our project, we're essentially attempting to do functional testing of all BELs inside the Versal slices. One of the requirements for this project is that we instantiate all possible BELs and test their I/O pins. This way, we can provide a level of assurance for users of Versal that their particular device is free from defects/faults that might arise over the lifecycle of the device. 

Direct instantiation of the IMRs would be ideal, but since this is not supported I have been trying to research other methods to get Vivado to instantaiate them. The documentation does not seem to provide any guidance on what types of designs would cause these resources to be used.

Do you have any ideas on how this could be done?

**Reply by clavin-xlnx** [reply]
Let me look into this further and I will get back to you.  In the meantime, would you mind sharing the target device you want to use?

**Reply by oshears** [reply]
Thanks Chris! I would appreciate any information you are able to find! The device that I am targeting for this project is `xcvp1202-vsva2785-2mp-e-s`

**Reply by clavin-xlnx** [reply]
The next version of Vivado will improve support for the IMR registers and should be coming out soon.  Let's wait until the release is made and see if it can answer your question.

**Reply by oshears** [reply]
Thanks for the update! I'll keep an eye out for the new release and hope it helps with this.

**Reply by clavin-xlnx** [reply, marked as answer]
Vivado 2023.2 is available now (https://www.xilinx.com/support/download.html), try using a LOC constraint in Vivado to place the IMR.

**Reply by oshears** [reply]
Looks like 2023.2 supports device image creation with IMRs. Thanks for your help!
